File: algorithm/20-Valid-Parentheses/isValid.py
# 一刷时间: 2019
# 链接：https: // leetcode-cn.com/problems/valid-parentheses

class Solution:
    def isValid(self, s):
        # 简洁但是效率低
        # while '{}' in s or '()' in s or '[]' in s:
        #     s = s.replace('{}', '')
        #     s = s.replace('[]', '')
        #     s = s.replace('()', '')
        # return s == ''
        stack = []  # 初始化栈
        mapping = {')': '(', ']': '[', '}': '{'}  # 字典存储所有的匹配类型
        for char in s:  # 遍历字符串
            if (char in mapping):  # 当前字符为闭合符号时
                top_element = stack.pop() if stack else '#'  # 栈顶元素出栈
                if mapping[char] != top_element:  # 没有匹配的类型，返回 false
                    return False
            else:
                stack.append(char)  # 开括号存储到栈里
        return not stack

    def __init__(self, arr):
        for item in arr:
            print(self.isValid(item))


arr = ['()', '()[]{}', "(]", "([)]", "{[]}", '']
Solution(arr)

File: algorithm/20-Valid-Parentheses/isValid.js
//  一刷时间: 2019
//  链接：https: // leetcode-cn.com/problems/valid-parentheses

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let valid = true;
  let stack = [];
  let map = {
    '(': ')',
    '[': ']',
    '{': '}'
  };
  const F = Object.keys(map);

  for (let i in s) {
    let value = s[i];

    const outStack = () => {
      const peak = stack.pop();
      if (value !== map[peak]) {
        valid = false;
      }
    };
    F.indexOf(value) !== -1 ? stack.push(value) : outStack();
  }
  if (stack.length > 0) return false;
  return valid;
};
